### ISSUE TITLE
How to best deal with corrections with regards to train/validation/test split

### ISSUE THREAD

**Issue by jf99** [mozilla/voice-web#1049]
According to #806 the data will be split into disjunct sets, so that
- each reader is in no more than one set and
- each sentence is in no more than one set (focus of this issue).

I don't know how this is currently implemented, but it might be worth to spend some thoughts on it. Essentially, I see two possible approaches:
- 1.) Split the sentences based on line number. E.g. numbers ending on 9 are test set, ending on 8 is validation set and the like.
- 2.) Compute a hash value out of the sentence string and use it to separate the sets.

The choice for either one of the possibilities has consequences when it comes to editing the sentence database. Obviously, despite all proofreading, belated corrections will be necessary here and there. If you choose possibility 1, then it must be prohibited to insert or delete lines before existing ones. Otherwise the following sentences would move to another dataset.

On the other hand, if you choose possibility 2, even a minor correction (e.g. inserting a comma) might assign the corrected sentence to a new dataset. As a result, you would have the (almost) same sentences in two different datasets. To get around this problem, one could delete all recordings based on the uncorrected sentence. However, a drawback of this measure would be that even minor corrections could destroy a hundred of valuable recordings. So at least the corrector should be warned.

Is there a better solution than the proposed?

**Comment by mikehenrty** [comment]
@Gregoor When we do a sentence correction (ie. modify a sentence line), what happens to the old sentence? Is it deleted from the DB?

**Comment by Gregoor** [comment]
@mikehenrty only unused sentences are deleted. Hence, if we do a correction of a sentence for which there is a clip, it'll keep existing and a new sentence will be added (which could indeed put that sentence into a different bucket).

**Comment by Gregoor** [comment]
@jf99 if the sentence doesn't change, it'll stay in the DB (basically untouched).

**Comment by mikehenrty** [comment]
> Hence, if we do a correction of a sentence for which there is a clip, it'll keep existing and a new sentence will be added

@Gregoor, will the existing sentence be marked as "inactive" so that we no longer record new versions of it?

**Comment by mikehenrty** [comment]
> new sentence will be added (which could indeed put that sentence into a different bucket).

This is not ideal, but fine I think. We have lots of sentences that are *close* phonetically, but we still bucket randomly, so unless we implement some sort of bucketing heuristics around phoneneme's, we'll take some "close" sentences in different buckets (so long as there are not too many).

**Comment by Gregoor** [comment]
> @Gregoor, will the existing sentence be marked as "inactive" so that we no longer record new versions of it?

yup. our current sentence import process (which happens on every startup) looks like this:
1. delete all unused sentences from DB
2. flag the remaining ones as inactive
3. insert all sentences from the files which aren't in the DB / mark the ones that are in the files as active.

**Comment by mikehenrty** [comment]
Ok, thanks for bringing up this issue @jf99. Due to the difficulty of separating by phonemes automatically, we will allow sentence corrections to be randomly to buckets based on existing criteria (with the old sentences being removed). Let's move forward with the corrections in your PRs.

**Comment by drzraf** [comment]
Orphaning voices when a small, non-semantic, presentation-only change occurs is a terrible loss of data.

Eg: #1191 change the presentation of 4 lines (many more PR exists doing similar thing).

One way to deal with this on server side would be to reassign voice from the old line to the new one **assuming lines weren't moved!**.

Apparently, hash is generated [that way](https://github.com/mozilla/voice-web/blob/288812bc6a9bfef0fdca636eaaf73c9868514231/scripts/extractSentences.js#L5):
`crypto.createHash("sha256").update(data).digest("base64")`

So, in order to change the voice's assignation for a given commit :
```
curl -s https://patch-diff.githubusercontent.com/raw/mozilla/voice-web/pull/1191.diff | \
  sed -ne '/^[+-]/s/.//p' | tail -n+3 | while read f; \
      do \
         echo -e "$f:\t$(sha256sum <<<"$f" | awk '{print $1}' | base64 -)"; \
      done
```

Brings:
```
elles est partie.:	NWYzNzAwZDk0MGEzM2VjNmJjZDYwNWVkYzQ2NGRjNmQ0MzM1OTMzODQ4OGM3MDRmNzQ4NjgzZmVl
M2Q0OTY2MAo=
elle est partie.:	YTRlZDAwMDE3ZTYyYTA1ZjAxMWNjMDExM2I3OTk0ZTIwZjA2MTdkMjExMDk1MzU5ZWZmZmI3NjBj
```
